ibm/resource_ibm_cloudant.go

```diff
@@ -0,0 +1,603 @@
+	if err != nil {
+		return err
+	}
+
+	// if matches an instance creation default skip request
+	if d.Get("include_data_events").(bool) {
+		err := updateCloudantActivityTrackerEvents(client, d)
+		if err != nil {
+			return fmt.Errorf("Error updating activity tracker events: %s", err)
+		}
+	}
+
+	// if matches an instance creation default skip request
+	if d.Get("capacity").(int) > 1 {
+		err := updateCloudantInstanceCapacity(client, d)
+		if err != nil {
+			return fmt.Errorf("Error retrieving capacity throughput information: %s", err)
+		}
+	}
+
+	err = updateCloudantInstanceCors(client, d)
+	if err != nil {
+		return fmt.Errorf("Error updating CORS settings: %s", err)
+	}
+
+	return resourceIBMCloudantRead(d, meta)
+}
+
+func resourceIBMCloudantRead(d *schema.ResourceData, meta interface{}) error {
+	err := resourceIBMResourceInstanceRead(d, meta)
+	if err != nil {
+		return err
+	}
+
+	err = setCloudantResourceControllerURL(d, meta)
+	if err != nil {
+		return err
+	}
+
+	client, err := getCloudantClient(d, meta)
+	if err != nil {
+		return err
+	}
+
+	// if resource was imported, set missing legacy_credentials
+	// to default false since we don't have this value exposed on the broker
+	if _, ok := d.GetOkExists("legacy_credentials"); !ok {
+		d.Set("legacy_credentials", false)
+	}
+
+	err = setCloudantActivityTrackerEvents(client, d)
+	if err != nil {
+		return err
+	}
+
+	err = setCloudantInstanceCapacity(client, d)
+	if err != nil {
+		return err
+	}
+
+	err = setCloudantInstanceCors(client, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceIBMCloudantUpdate(d *schema.ResourceData, meta interface{}) error {
+	d.Set("service", "cloudantnosqldb")
+
+	err := validateCloudantInstanceCapacity(d)
+	if err != nil {
+		return err
+	}
+
+	err = validateCloudantInstanceCors(d)
+	if err != nil {
+		return err
+	}
+
+	err = resourceIBMResourceInstanceUpdate(d, meta)
+	if err != nil {
+		return err
+	}
+
+	client, err := getCloudantClient(d, meta)
+	if err != nil {
+		return err
+	}
+
+	if d.HasChange("include_data_events") {
+		err := updateCloudantActivityTrackerEvents(client, d)
+		if err != nil {
+			return fmt.Errorf("Error updating activity tracker events: %s", err)
+		}
+	}
+
+	if d.HasChange("capacity") {
+		err := updateCloudantInstanceCapacity(client, d)
+		if err != nil {
+			return fmt.Errorf("Error retrieving capacity throughput information: %s", err)
+		}
+	}
+
+	if d.HasChange("enable_cors") {
+		err := updateCloudantInstanceCors(client, d)
+		if err != nil {
+			return fmt.Errorf("Error updating CORS settings: %s", err)
+		}
+	}
+
+	return resourceIBMCloudantRead(d, meta)
+}
+
+func setCloudantResourceControllerURL(d *schema.ResourceData, meta interface{}) error {
+	crn := d.Get(ResourceCRN).(string)
+	rcontroller, err := getBaseController(meta)
+	if err != nil {
+		return err
+	}
+	d.Set(ResourceControllerURL, rcontroller+"/services/cloudantnosqldb/"+url.QueryEscape(crn))
```

Comment:
I tried to get this attribute from `resourcecontrollerv2` and `globaltaggingv1` APIs without much success. I'm not sure if it is exposed on any metadata we can use, since the rest of the providers doing the same as us, i.e. constructing it manually. It's not even clear to me _what_ is this attribute for, I've found slack discussion about it where it's mentioned that this parameter used for Schematics [slack ref](https://ibm-cloudplatform.slack.com/archives/C01P7QR03FC/p1616057759006800?thread_ts=1614284998.013700&cid=C01P7QR03FC)

I guess out options here either keep it as it is or remove from schema completely. I've tried it and it is possible to do. We might have to and add it back if there will be complains during upstream PR, but it is what it is.

@ricellis let me know what do you think.

Reply:
It looks like registry got the go ahead to remove the `ResourceControllerURL` - https://github.com/IBM-Cloud/terraform-provider-ibm/pull/2372/files#diff-20694cf456cfa8aabacde5afd133af63dea8a5e8ea33fbb71e6d19f3bb71a6c0L54 - so it seems like it is not a requirement.
I think I'd rather we drop it and only add it back if requested for a specific purpose than leave something potentailly fragile around for no [apparent] reason.

Reply:
This is fairly confusing, but while removing `ResourceControllerURL` from schema works for me when I run terraform manually it crashes in tests, because we are using `IBMResourceInstanceRead` functions and they are trying to set this attribute while it's not on our schema.

I'm not sure why it is working when I'm running things manually, but it seems that we stuck with it. Registry got away with removing it because they are not delegating resource creation/reading to other modules. I'm going to keep it as it is, can't remove key from `ResourceData` once it set, so we can't hide it after `ResourceInstance` set it.

Reply:
acknowledged